packages/core/src/directory.ts

```diff
@@ -0,0 +1,107 @@
+    userUpdated(user: DirectoryUser, orgId: OrgID): Promise<void>;
+    userDeleted(user: DirectoryUser, orgId: OrgID): Promise<void>;
+
+    groupCreated(group: DirectoryGroup, orgId: OrgID): Promise<void>;
+    groupUpdated(group: DirectoryGroup, orgId: OrgID): Promise<void>;
+    groupDeleted(group: DirectoryGroup, orgId: OrgID): Promise<void>;
+}
+
+export interface DirectoryProvider {
+    subscribe(sub: DirectorySubscriber): void;
+}
+
+export class DirectorySync implements DirectorySubscriber {
+    constructor(public readonly server: Server, providers: DirectoryProvider[] = []) {
+        for (const provider of providers) {
+            provider.subscribe(this);
+        }
+    }
+
+    async userCreated(user: DirectoryUser, orgId: string) {
+        const org = (orgId && (await this.server.storage.get(Org, orgId))) || null;
+        if (org && org.directory.syncProvider === "scim" && org.directory.syncMembers) {
+            const memberExists = org.members.some((member) => member.directoryExternalId === user.externalId);
+
+            if (memberExists) {
+                return;
+            }
+
+            org.members.push(
+                new OrgMember({
+                    directoryExternalId: user.externalId,
+                    name: user.name,
+                    email: user.email,
+                    status: OrgMemberStatus.Provisioned,
+                    updated: new Date(),
+                })
+            );
+
+            await this.server.updateMetaData(org);
+            await this.server.storage.save(org);
+        }
+    }
+
+    async userUpdated(user: DirectoryUser, orgId: string): Promise<void> {
+        const org = (orgId && (await this.server.storage.get(Org, orgId))) || null;
+        if (org && org.directory.syncProvider === "scim" && org.directory.syncMembers) {
+            const existingUser = org.members.find((member) => member.directoryExternalId === user.externalId);
+
+            if (!existingUser) {
+                return;
+            }
+
+            existingUser.name = user.name;
+            existingUser.updated = new Date();
+
+            await this.server.updateMetaData(org);
+            await this.server.storage.save(org);
+        }
+    }
+    async userDeleted(user: DirectoryUser, orgId: string): Promise<void> {
+        const org = (orgId && (await this.server.storage.get(Org, orgId))) || null;
+        if (org && org.directory.syncProvider === "scim" && org.directory.syncMembers) {
+            const existingUser = org.members.find((member) => member.directoryExternalId === user.externalId);
+
+            if (!existingUser) {
+                return;
+            }
+
+            await org.removeMember(existingUser, false);
+
+            // Remove any existing invites
+            const existingInvite = org.invites.find((invite) => invite.email === existingUser.email);
+            if (existingInvite) {
+                org.removeInvite(existingInvite);
+            }
```

Comment:
@MaKleSoft this was enough for me to get it working locally. Let me know if there's anything else I'm missing.

Reply:
Unfortunately there is a little bit more work involved. See https://github.com/padloc/padloc/blob/v4/packages/core/src/server.ts#L1115-L1133